tests/pairlist/test_pairlist.py

```diff
@@ -701,11 +718,61 @@ def test_pricefilter_desc(mocker, whitelist_conf, markets, pairlistconfig,
             freqtrade = get_patched_freqtradebot(mocker, whitelist_conf)
 
 
-def test_pairlistmanager_no_pairlist(mocker, markets, whitelist_conf, caplog):
+def test_pairlistmanager_no_pairlist(mocker, whitelist_conf):
     mocker.patch('freqtrade.exchange.Exchange.exchange_has', MagicMock(return_value=True))
 
     whitelist_conf['pairlists'] = []
 
     with pytest.raises(OperationalException,
                        match=r"No Pairlist Handlers defined"):
         get_patched_freqtradebot(mocker, whitelist_conf)
+
+
+@pytest.mark.parametrize("pairlists,pair_allowlist,overall_performance,allowlist_result", [
+    # Happy path, descending order, all values filled
+    ([{"method": "StaticPairList"}, {"method": "PerformanceFilter"}],
+     ['ETH/BTC', 'TKN/BTC'],
+     [{'pair': 'TKN/BTC', 'profit': 5, 'count': 3}, {'pair': 'ETH/BTC', 'profit': 4, 'count': 2}],
+     ['TKN/BTC', 'ETH/BTC']),
+    # Performance data outside allow list ignored
+    ([{"method": "StaticPairList"}, {"method": "PerformanceFilter"}],
+     ['ETH/BTC', 'TKN/BTC'],
+     [{'pair': 'OTHER/BTC', 'profit': 5, 'count': 3},
+      {'pair': 'ETH/BTC', 'profit': 4, 'count': 2}],
+     ['ETH/BTC', 'TKN/BTC']),
+    # Partial performance data missing and sorted between positive and negative profit
+    ([{"method": "StaticPairList"}, {"method": "PerformanceFilter"}],
+     ['ETH/BTC', 'TKN/BTC', 'LTC/BTC'],
+     [{'pair': 'ETH/BTC', 'profit': -5, 'count': 100},
+      {'pair': 'TKN/BTC', 'profit': 4, 'count': 2}],
+     ['TKN/BTC', 'LTC/BTC', 'ETH/BTC']),
+    # Tie in performance data broken by count
+    ([{"method": "StaticPairList"}, {"method": "PerformanceFilter"}],
+     ['ETH/BTC', 'TKN/BTC', 'LTC/BTC'],
+     [{'pair': 'LTC/BTC', 'profit': -5, 'count': 101},
+      {'pair': 'TKN/BTC', 'profit': -5, 'count': 2},
+      {'pair': 'ETH/BTC', 'profit': -5, 'count': 100}],
+     ['LTC/BTC', 'ETH/BTC', 'TKN/BTC']),
+])
+def test_performance_filter(mocker, whitelist_conf, pairlists, pair_allowlist, overall_performance,
```

Comment:
Agreed that it should be more explicitly tested. 

[This comment](https://github.com/freqtrade/freqtrade/issues/3441#issuecomment-735045902) alludes to some of the current strange sorting that occurs when no performance data is available - seemingly sorted by length of string.

Reply:
As it was, I was seeing a `KeyError: 'pair'` when providing an empty list. So I modified PerformanceFilter to skip performance-based processing if that case.

Reply:
Maybe doing the sorting based on `sort_values(by=['profit', 'count', 'pair'])`  would help to not sort by "length of the string" - but alphabetically (better than length of string i guess).

Either way, skipping in case of no performance data available is perfectly fine - i think that was discussed in the issue too - as this filter will have a undefined startup-period anyway.

Reply:
Resolving this in deference to [comment](https://github.com/freqtrade/freqtrade/pull/4009#discussion_r532086775) at location of remaining related test.